resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -750,20 +799,27 @@ public void waitForAutoBuildOff() {
 	}
 
 	private static void waitFor(Job job) {
+		IJobManager jobManager = Job.getJobManager();
+
 		// Need to loop because jobs can reschedule itself and concurrent running
 		// background jobs change the states too
-		while (!(job.getState() == Job.NONE)) {
+		while (job.getState() != Job.NONE) {
 			// Need to wake up thread to finish as soon as possible:
-			while (!(job.getState() == Job.RUNNING || job.getState() == Job.NONE)) {
-				Job.getJobManager().wakeUp(ResourcesPlugin.FAMILY_AUTO_BUILD);
+			while (job.getState() != Job.RUNNING && job.getState() != Job.NONE) {
+				jobManager.wakeUp(ResourcesPlugin.FAMILY_AUTO_BUILD);
 				Thread.yield();
 				// After wakeup the woken job may go into sleep again when asynchronous
 				// workspace save interrupts autobuild so we need to wait till RUNNING or NONE
 				// (happens after each JUnit class)
+
+				if (jobManager.isSuspended() && job.getState() != Job.RUNNING && job.getState() != Job.NONE)
```

Comment:
The checks for the state should not be necessary anymore as the loop condition already checked them. In addition: couldn't the condition be checked earlier as waking up jobs and yielding this thread will have not effect in case the job manager is suspended anyway.